nx/source/services/avm.c

```diff
@@ -0,0 +1,109 @@
+#include "service_guard.h"
+#include "services/avm.h"
+#include "runtime/hosversion.h"
+
+static Service g_AvmSrv;
+
+NX_GENERATE_SERVICE_GUARD(avm);
+
+Result _avmInitialize(void) {
+    if (hosversionBefore(6,0,0))
+        return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+
+    Result rc=0;
+
+    rc = smGetService(&g_AvmSrv, "avm");
+
+    if (R_SUCCEEDED(rc)) rc = serviceConvertToDomain(&g_AvmSrv);
+
+    return rc;
+}
+
+void _avmCleanup(void) {
+    serviceClose(&g_AvmSrv);
+}
+
+Service *avmGetServiceSession(void) {
+    return &g_AvmSrv;
+}
+
+static Result _avmGetVersion(u32 cmd_id, u64 id_1, u64 id_2, u32 *version) {
+    const struct {
+        u64 a, b;
+    } in = { id_1, id_2 };
+
+    return serviceDispatchInOut(&g_AvmSrv, 100, in, *version);
+}
+
+Result avmGetHighestAvailableVersion(u64 id_1, u64 id_2, u32 *version) {
+    return _avmGetVersion(100, id_1, id_2, version);
+}
+
+Result avmGetHighestRequiredVersion(u64 id_1, u64 id_2, u32 *version) {
+    return _avmGetVersion(101, id_1, id_2, version);
+}
+
+Result avmGetVersionListEntry(u64 application_id, AvmVersionListEntry *entry) {
+    return serviceDispatchInOut(&g_AvmSrv, 102, application_id, *entry);
+}
+
+Result avmGetVersionListImporter(AvmVersionListImporter *out) {
+    return serviceDispatch(&g_AvmSrv, 103,
+        .out_num_objects = 1,
+        .out_objects = &out->s,
+    );
+}
+
+Result avmGetLaunchRequiredVersion(u64 application_id, u32 *version) {
+    return serviceDispatchInOut(&g_AvmSrv, 200, application_id, *version);
+}
+
+static Result _avmPushVersion(u32 cmd_id, u64 application_id, u32 version) {
+    const struct {
+        u64 application_id;
```

Comment:
Is this param order really correct?

Reply:
whoops... happened while merging commands ^^